### PR TITLE
chore: Apply rounded courners to first and last stacked bars only

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

Updated border rounding for stacked bar charts (horizontal and vertical, classic and VR) so that not every segment but only the edge ones are rounded from the top or from the bottom. That is to avoid an issue when small segments that are a few pixel tall/wide look awkward when rounded.

**Before**
Screenshot 2024-04-18 at 15 21 25

**After**
Screenshot 2024-04-18 at 15 23 03

### How has this been tested?

* Existing permutations for bar charts

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/mixed-line-bar-chart/bar-series.tsx [hunk 1]
```diff
@@ -126,30 +135,68 @@ export default function BarSeries<T extends ChartDataTypes>({
...
   );
 }
+
+function stretchRect(
+  rect: { x: number; y: number; height: number; width: number },
+  direction: 'left' | 'right' | 'up' | 'down'
+) {
+  switch (direction) {
+    case 'up':
+      return { ...rect, height: rect.height + 10 };
```
Comment (just-boris) [inline src/mixed-line-bar-chart/bar-series.tsx:194]: Where are these 10 numbers came from? I do not see it in the original code

File: src/mixed-line-bar-chart/data-series.tsx [hunk 2]
```diff
@@ -65,6 +65,13 @@ export default function DataSeries<T extends ChartDataTypes>({
     return calculateOffsetMaps(barData);
   }, [visibleSeries, stackedBars]);
 
+  const getStackedMinimum = (xValue: ChartDataTypes) => {
+    return stackedBarOffsetMaps?.minValues.get(getKeyValue(xValue)) ?? 0;
```
Comment (just-boris) [inline src/mixed-line-bar-chart/data-series.tsx:69]: When is this value undefined and why is it safe to assume 0 here?

File: src/mixed-line-bar-chart/utils.ts [hunk 3]
```diff
@@ -95,9 +109,13 @@ export function calculateOffsetMaps(
...
         const lastValue = lastMap?.negativeOffsets.get(key) || 0;
         map.negativeOffsets.set(key, lastValue + y);
+        setMinValue(key, lastValue + y);
+        setMaxValue(key, lastValue + y);
       } else {
         const lastValue = lastMap?.positiveOffsets.get(key) || 0;
         map.positiveOffsets.set(key, lastValue + y);
+        setMinValue(key, lastValue + y);
+        setMaxValue(key, lastValue + y);
       }
```
Comment (just-boris) [inline src/mixed-line-bar-chart/utils.ts:119]: Don't you see any issues when repeating `lastValue + y` 6 times here?

File: src/mixed-line-bar-chart/utils.ts [hunk 4]
```diff
@@ -108,6 +126,8 @@ export function calculateOffsetMaps(
 
     return acc;
   }, [] as StackedOffsets[]);
+
+  return { minValues, maxValues, seriesOffsets };
```
Comment (just-boris) [inline src/mixed-line-bar-chart/utils.ts:130]: If `seriesOffsets` already contains sorted data, isn't it self explanatory which item is min and which item is max? Do you really need a separate value here?

Comment (pan-kot) [reply src/mixed-line-bar-chart/utils.ts:130]: The `seriesOffsets` includes data for a single series only. For example:

```
series = "John"
seriesOffsets = {
  positive: { "Apples": 0, "Oranges": 5 },
  negative: { "Pears": 0  }
}
```

The chart is rendered series by series, so the John series will render right above the X axis for "Apples", above the X axis and some other series for "Oranges", and right below the X axis for "Pears". This carries no information about other series.

File: src/mixed-line-bar-chart/utils.ts (hunk 3 above)
Comment (pan-kot) [reply src/mixed-line-bar-chart/utils.ts:119]: I will refactor this utility, resolving the thread.

File: src/mixed-line-bar-chart/data-series.tsx (hunk 2 above)
Comment (pan-kot) [reply src/mixed-line-bar-chart/data-series.tsx:69]: Logically, the min/max values are defined for each value so do not expect any to be undefined. However, I found how to make the overall code simpler - resolving the thread.

File: src/mixed-line-bar-chart/bar-series.tsx (hunk 1 above)
Comment (pan-kot) [reply src/mixed-line-bar-chart/bar-series.tsx:194]: That's a good question. I will add a code comment to explain that.

File: src/mixed-line-bar-chart/utils.ts (hunk 4 above)
Comment (pan-kot) [reply src/mixed-line-bar-chart/utils.ts:130]: Thanks for the comment though, looking at it now I realised it worth taking a step back and create a better data structure than series offsets to accommodate all needs.

File: src/mixed-line-bar-chart/bar-series.tsx (hunk 1 above)
Comment (just-boris) [reply src/mixed-line-bar-chart/bar-series.tsx:199]: See the comment now but still do not understand what is happening? Why is this 10 pixels and not 5 or 100?

File: src/mixed-line-bar-chart/data-series.tsx [hunk 5]
```diff
@@ -62,7 +61,7 @@ export default function DataSeries<T extends ChartDataTypes>({
         barData.push([]);
       }
     });
-    return calculateOffsetMaps(barData);
+    return calculateStackedBarValues([...barData]);
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/mixed-line-bar-chart/data-series.tsx]: Sorry, that's a leftover code. I was experimenting with changing the argument type to accept the readonly array. Reverted this back.

**Pull request merged**